addons/udes_stock/models/stock_location.py

```diff
@@ -134,6 +135,50 @@ def _domain_height_category(self):
         see u_countable field.""",
     )
 
+    u_heatmap_data_updated = PreciseDatetime(
+        string="Heatmap Data Updated",
+        help="The date when the location name, co-ordinates or size was last changed.",
+        default=PreciseDatetime.now(),
+    )
+
+    def set_u_heatmap_data_updated(self, vals):
+        """
+        Iterate over locations in `self`, checking if any heatmap fields
+        have been written to with updated values, and if so - dispatch to write()
+        with a bypass in context (to avoid recursion) to update `u_heatmap_data_updated`
+        with the current date & time.
+
+        :param: vals: dict() of vals from `write()`
+        """
+        StockLocation = self.env["stock.location"]
+        locations_to_update = StockLocation.browse()
+        for location in self:
+            heatmap_fields = ["u_size", "name", "posx", "posy", "posz"]
+            # Don't need to check old/new values of these fields if they aren't in vals
+            if not any([x in vals.keys() for x in heatmap_fields]):
```

Comment:
We can put this (along with `heatmap_fields`) outside of the loop completely. Even if there are multiple locations in `self`, `vals` will always be the same for each location.

So we can check the vals before we start looping through locations.